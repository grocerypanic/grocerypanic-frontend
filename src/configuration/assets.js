const Assets = {
  favicon: `${process.env.PUBLIC_URL}/favicon.ico`,
  nonBreakingSpace: "\u00A0",
  cart: `${process.env.PUBLIC_URL}/images/cart_desktop.jpg`,
  bill: `${process.env.PUBLIC_URL}/images/bill_desktop.jpg`,
  cart_mobile: `${process.env.PUBLIC_URL}/images/cart_mobile.jpg`,
  bill_mobile: `${process.env.PUBLIC_URL}/images/bill_mobile.jpg`,
};

export default Assets;
