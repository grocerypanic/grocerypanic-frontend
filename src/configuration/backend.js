export const SafeMethods = ["GET", "HEAD", "OPTIONS", "TRACE"];
export const ItemFilters = ["preferred_stores", "shelf"];
export const FilterTag = "filter_tag";

export const Constants = {
  csrfLocalStorage: "csrf-token",
  csrfHeaderName: "X-CSRFToken",
  genericAPIError: "API Error",
  csrfErrorMessage: "Refresh csrf and try again.",
  alreadyRegistered: "User is already registered with this e-mail address.",
  minimumQuanity: 0,
  maximumQuantity: 10000,
  minimumShelfLife: 1,
  maximumShelfLife: 365 * 3,
  defaultShelfLife: 7,
  minimumPrice: 0,
  maximumPrice: 9999,
  maximumTransactionHistory: 14,
  AnalyticsCookieName: "PanicDataConsent-" + process.env.NODE_ENV.toUpperCase(),
  duplicateObjectApiError: { non_field_errors: ["Object already exists."] },
  pageLookupParam: "page",
  pageOverrideParam: "all_results",
};

export const Providers = {
  google: "google",
  facebook: "facebook",
};

export const Paths = {
  googleLogin: "/api/v1/auth/social/google/",
  facebookLogin: "/api/v1/auth/social/facebook/",
  manageShelves: "/api/v1/shelves/",
  manageStores: "/api/v1/stores/",
  manageItems: "/api/v1/items/",
  refreshCSRF: "/api/v1/auth/csrf/",
  manageTransactions: "/api/v1/transactions/",
};

const month = 30;
const year = 365;

export const ShelfLifeConstants = [
  { name: "3 Days", id: 3 },
  { name: "1 Week", id: 7 },
  { name: "2 Weeks", id: 14 },
  { name: "3 Weeks", id: 21 },
  { name: "1 Month", id: month },
  { name: "2 Months", id: month * 2 },
  { name: "3 Months", id: month * 3 },
  { name: "6 Months", id: month * 6 },
  { name: "1 Year", id: year },
  { name: "2 Years", id: year * 2 },
  { name: "3 Years", id: year * 3 },
];
