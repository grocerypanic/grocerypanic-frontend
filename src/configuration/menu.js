import Routes from "./routes";

const Options = [
  { name: "All Items", location: Routes.items },
  { name: "Shelves", location: Routes.shelves },
  { name: "Stores", location: Routes.stores },
];

export default Options;
