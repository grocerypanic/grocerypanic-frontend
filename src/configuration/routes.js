const Routes = {
  root: "/",
  menu: "/menu",
  about: "/about",
  privacy: "/privacy",
  signin: "/signin",
  shelves: "/shelves",
  splash: "/",
  stores: "/stores",
  items: "/items",
  create: "/create",
  details: "/details/:id",
  goBack: "back",
};

export const External = {
  feedback: "https://www.sharedvisionsolutions.com/#contact",
  credit: "https://www.sharedvisionsolutions.com/",
};

export default Routes;
