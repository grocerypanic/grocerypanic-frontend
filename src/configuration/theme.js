export const primary = "#006600";
export const white = "#FFFFFF";
export const secondary = "#868e96";
export const tertiary = "#808080";
export const error = "#FF0000";
export const regular = "#000000";
export const success = "#5cb85c";
export const highlight = "blue";
export const selected = "red";
export const danger = "#d9534f";
export const dark = "#495057";
export const black = "#000000";
export const warning = "#ffc107";

export const ui = {
  alertTimeout: 2000,
  resizeTimeout: 1,
};

export const navbarSize = "50px";
export const tabSize = "26px";
export const itemAttributes = {
  height: 55,
  padding: 5,
  border: 10,
  spacing: 5,
  innerHeight: 40,
  placeHolderPadding: 12,
};

export const headerMobileThreshold = 480;
export const mobileThreshold = 480;

export const graph = {
  quantityLine: "rgb(2, 117, 216)",
  changeLine: "rgb(92, 184, 92)",
  topPadding: -12,
  holdingPatternHeight: "134px",
  chartContainerHeight: "150px",
  holdingPatternColor: "secondary",
  holdingPatternScale: 1,
  holdingPatternDivHeight: 20,
  holdingPatternAnimation: "grow",
};
