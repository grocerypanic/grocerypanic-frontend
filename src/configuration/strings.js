import assets from "./assets";

export const Strings = {
  en: {
    SignIn: {
      Title: "Sign In",
      LoginMessageGoogle: "Login With Google",
      LoginMessageFacebook: "Login With Facebook",
      ErrorLoginFailure: "Unable to authenticate!\nPlease try again.",
      ErrorAuthExpired: "Authentication Expired!\nPlease login again.",
      ErrorDuplicateAccount:
        "You have already signed up!\nTry logging in with another social account.",
      PendingSocialConnection: "Connecting . . .",
    },
    ErrorDialogue: {
      ErrorDialogueConfirm: "OK",
      ErrorDialogueTitle: "Error!",
    },
    Copyight: {
      CopyrightDeclaration: "Copyright Â© ",
      CopyrightMessage: "Your Website",
    },
    App: {
      Suspense: "Suspense filled message...",
    },
    CookiePolicy: {
      CookieMessage:
        "This website would like to use cookies to measure how you as an individual interact with it.  We'll use this information to improve the service.",
      CookieAcceptText: "I'm ok with that",
      CookieDeclineText: "I Decline",
    },
    MainMenu: {
      Title: "Main Menu",
      HeaderTitle: "Panic: Main Menu",
      HelpText: "Make a selection to proceed.",
    },
    ItemDetails: {
      Title: "Edit Item Details",
      HeaderTitle: "Details",
      HelpText:
        "Fill out the values, click the save button to finalize.\nDon't modify quantity here.",
      NameLabel: "Name",
      QuantityLabel: "Qty",
      PriceLabel: "Price",
      StoresLabel: "Stores",
      SaveButton: `${assets.nonBreakingSpace}Save${assets.nonBreakingSpace}`,
      DeleteButton: "Delete",
      MultiSelectHelp: "",
      ShelvesDetail: "Where do you keep this at home?",
      ShelfLifeDetail: "How long does this item typically keep?",
      PerferredLocationDetails: "Where do you prefer to buy this?",
      ErrorUnselectedStore: "Choose a preferred stored.",
      SaveAction: "Saved",
      DeleteAction: "Deleting ... :(",
      Tabs: {
        Edit: "Edit",
        Stats: "Stats",
      },
      ApiError:
        "Unable to retrieve data from the API!\nDoes this content really exist?\nRetry, or change your query.",
    },
    ItemList: {
      ApiError:
        "Unable to retrieve data from the API!\nDoes this content really exist?\nRetry, or change your query.",
    },
    InventoryPage: {
      Title: "Inventory (All)",
      HeaderTitle: "Inventory",
      HelpText:
        "Click and hold the item's name for details.\nUse the create button (upper right) to add new items.\nIncrese or decrease quantity with the buttons.",
      Save: "save",
      Delete: "delete",
      PlaceHolderMessage: "You have no items yet.",
      ErrorInsufficientInventory: "You don't have that many.",
      Quantity: {
        Title: "Quantity",
        Message: "This is how many items you have in stock.",
      },
      Expired: {
        Title: "Expired",
        Message:
          "This is how many items you have in stock that may be expired.",
      },
    },
    SimpleList: {
      ValidationAlreadyExists: "This already exists",
      ValidationFailure: "Enter a valid name",
      SaveButton: "save",
      DeleteButton: "delete",
      CreatedAction: "Created",
      DeletedAction: "Deleted",
    },
    ShelfPage: {
      Title: "Your Shelves",
      HeaderTitle: "Shelves",
      PlaceHolderMessage: "You Have No Shelves Yet.",
      HelpText:
        "Click a shelf to select, click and hold to delete.\nUse the create button (upper right) to add new shelves.",
    },
    StorePage: {
      Title: "Your Stores",
      HeaderTitle: "Stores",
      PlaceHolderMessage: "You Have No Stores Yet.",
      HelpText:
        "Click a store to select, click and hold to delete.\nUse the create button (upper right) to add new stores.",
    },
    Testing: {
      GenericTranslationTestString: "GenericTranslationTestString",
      GenericMultiLineTranslationTestString:
        "GenericTranslationTestString\nGenericTranslationTestString",
    },
    PlaceHolder: {
      PlaceHolderMessage: "I'm only a placeholder.",
    },
  },
};

// Export english as default for matching during tests
export default Strings.en;
