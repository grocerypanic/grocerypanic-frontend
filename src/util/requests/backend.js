import { Constants, SafeMethods } from "../../configuration/backend";
import debug from "../debug";

const Backend = (method, path, data = null) => {
  // Prepend Backend Hostname and Protocol If Needed
  let url;
  if (!path.includes(process.env.REACT_APP_PANIC_BACKEND)) {
    url = `${process.env.REACT_APP_PANIC_BACKEND}${path}`;
  } else {
    url = path;
  }

  debug(`API ${method}:\n ${url}`);
  if (data) debug(`Body:\n ${JSON.stringify(data)}`);
  let csrf;
  let statusCode;

  const options = {
    method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    credentials: "include",
  };

  if (data) options.body = JSON.stringify(data);
  if (!SafeMethods.includes(method)) {
    csrf = localStorage.getItem(Constants.csrfLocalStorage);
    if (csrf) options.headers[Constants.csrfHeaderName] = csrf;
  }

  const response = fetch(url, options)
    .then(async (fetchResponse) => {
      const contentType = fetchResponse.headers.get("content-type");
      statusCode = fetchResponse.status;
      if (contentType === null) return null;
      if (contentType.startsWith("application/json"))
        return await fetchResponse.json();
      return fetchResponse.text();
    })
    .then((fetchResponseDecoded) => {
      debug(`API Response Status Code:\n ${statusCode}`);
      if (fetchResponseDecoded) {
        debug(`API Response Data:\n`);
        debug(fetchResponseDecoded);
      }
      return [fetchResponseDecoded, statusCode];
    })
    .catch((err) => {
      return [Constants.genericAPIError, 500];
    });
  return response;
};

export default Backend;
