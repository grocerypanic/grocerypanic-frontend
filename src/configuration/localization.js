import i18n from "i18next";
import LanguageDetector from "i18next-browser-languagedetector";
import { Strings } from "./strings";

const i18nConfig = {
  fallbackLng: {
    default: ["en"],
  },
  debug: true,
  ns: ["translation"],
  defaultNS: "translation",
  keySeparator: ".",
  interpolation: {
    escapeValue: false,
    formatSeparator: ",",
  },
  react: {
    wait: true,
  },
};

i18n.use(LanguageDetector).init({
  resources: {
    ...Strings,
  },
  ...i18nConfig,
});

export default i18n;
