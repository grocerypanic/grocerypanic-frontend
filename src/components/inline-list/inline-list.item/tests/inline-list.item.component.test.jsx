import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import { FilterTag } from "../../../../configuration/backend";
import Routes from "../../../../configuration/routes";
import Strings from "../../../../configuration/strings";
import { highlight } from "../../../../configuration/theme";
import InlineListItem, { testIDs } from "../inline-list.item.component";

jest.mock("../../../../configuration/theme", () => {
  return {
    __esModule: true,
    ...jest.requireActual("../../../../configuration/theme"),
    ui: { alertTimeout: 10 },
  };
});

jest.mock("../../../../configuration/theme");

const initialProps = {
  item: { id: 0, name: "mockItem" },
  handleDelete: jest.fn(),
  history: { push: jest.fn() },
  objectClass: "shelf",
  transaction: false,
  redirectTag: "shelf",
  selected: null,
  setSelected: jest.fn(),
};

describe("InlineList", () => {
  let currentProps;
  let container;

  beforeEach(() => {
    jest.clearAllMocks();
    resetProps();
  });

  const resetProps = () => {
    currentProps = { ...initialProps };
  };

  const arrange = () => {
    const result = render(<InlineListItem {...currentProps} />);
    container = result.container;
  };

  describe("when outside of a transaction", () => {
    let item;
    let title;

    beforeEach(async () => {
      currentProps.transaction = false;
      currentProps.item.id = 1;
      arrange();
      item = await screen.findByTestId(testIDs.ListItemElement);
      title = await screen.findByTestId(testIDs.ListItemTitle);
    });

    it("should display the item's text in the correct colour", () => {
      expect(item).toHaveStyle({ color: highlight });
    });

    it("should display the item's title with correct text", () => {
      expect(title.innerHTML).toBe(currentProps.item.name);
    });

    it("should NOT display the input form", () => {
      expect(
        screen.queryByTestId(testIDs.ListItemNewItemInputElement)
      ).toBeNull();
    });

    it("should NOT display the save button", () => {
      expect(screen.queryByTestId(testIDs.ListItemSaveButton)).toBeNull();
    });

    it("should match the snapshot on file (styles)", () => {
      expect(container).toMatchSnapshot();
    });

    describe("when the title is clicked", () => {
      beforeEach(() => {
        expect(currentProps.history.push).toBeCalledTimes(0);
        fireEvent.click(title);
      });

      it("should push a route", async () => {
        await waitFor(() =>
          expect(currentProps.history.push).toBeCalledTimes(1)
        );
        const params = {};
        params[FilterTag] = currentProps.item.name;
        params[currentProps.redirectTag] = currentProps.item.id;
        params.class = currentProps.objectClass;
        const url = `${Routes.items}?${new URLSearchParams(params).toString()}`;
        expect(currentProps.history.push).toBeCalledWith(url);
      });
    });

    describe("when the item is clicked", () => {
      beforeEach(() => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
        fireEvent.mouseDown(item);
        fireEvent.mouseUp(item);
      });

      it("should set the new item to selected", async () => {
        await waitFor(() =>
          expect(currentProps.setSelected).toBeCalledTimes(1)
        );
        expect(currentProps.setSelected).toBeCalledWith(1);
      });
    });

    describe("when the item is long clicked", () => {
      let deleteButton;

      beforeEach(async () => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
        fireEvent.mouseDown(item);
        deleteButton = await screen.findByTestId(testIDs.ListItemDeleteButton);
      });

      it("should show the delete button with correct props", async () => {
        expect(deleteButton.innerHTML).toBe(Strings.InlineList.DeleteButton);
      });

      describe("when the delete button is pressed", () => {
        beforeEach(() => {
          expect(currentProps.handleDelete).toBeCalledTimes(0);
          fireEvent.click(deleteButton);
        });

        it("should call the deleteHandler with the item id", () => {
          expect(currentProps.handleDelete).toBeCalledTimes(1);
          expect(currentProps.handleDelete).toBeCalledWith(
            currentProps.item.id,
            currentProps.item.name
          );
        });
      });
    });
  });

  describe("when a transaction is ongoing", () => {
    let item;
    let title;

    beforeEach(async () => {
      currentProps.transaction = true;
      currentProps.item.id = 1;
      arrange();
      item = await screen.findByTestId(testIDs.ListItemElement);
      title = await screen.findByTestId(testIDs.ListItemTitle);
    });

    it("should display the item's text in the correct colour", () => {
      expect(item).toHaveStyle({ color: highlight });
    });

    it("should display the item's title with correct text", () => {
      expect(title.innerHTML).toBe(currentProps.item.name);
    });

    it("should NOT display the input form", () => {
      expect(
        screen.queryByTestId(testIDs.ListItemNewItemInputElement)
      ).toBeNull();
    });

    it("should NOT display the save button", () => {
      expect(screen.queryByTestId(testIDs.ListItemSaveButton)).toBeNull();
    });

    it("should match the snapshot on file (styles)", () => {
      expect(container).toMatchSnapshot();
    });

    describe("when the title is clicked", () => {
      beforeEach(() => {
        expect(currentProps.history.push).toBeCalledTimes(0);
        fireEvent.click(title);
      });

      it("should NOT push a route", () => {
        expect(currentProps.history.push).toBeCalledTimes(0);
      });
    });

    describe("when the item is clicked", () => {
      beforeEach(() => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
        fireEvent.mouseDown(item);
        fireEvent.mouseUp(item);
      });

      it("should set the new item to selected", () => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
      });
    });

    describe("when the item is long clicked", () => {
      beforeEach(() => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
        fireEvent.mouseDown(item);
      });

      it("should NOT show the delete button", (done) => {
        setTimeout(() => {
          expect(screen.queryByTestId(testIDs.ListItemDeleteButton)).toBeNull();
          done();
        }, 600);
      });
    });
  });
});
