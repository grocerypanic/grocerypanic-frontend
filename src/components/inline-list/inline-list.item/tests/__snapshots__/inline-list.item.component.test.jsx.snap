// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineList when a transaction is ongoing should match the snapshot on file (styles) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background: #FFFFFF;
  margin-bottom: 5px;
  padding: 5px;
  color: blue;
  min-width: 220px;
}

.c0 div:last-child {
  margin-left: auto;
}

.c0 .inline-list-item-title-div:hover {
  color: red;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 40px;
}

<div>
  <li
    class="c0"
    data-testid="ListItemElement"
  >
    <div
      class="c1 inline-list-item-title-div"
      data-testid="ListItemTitle"
    >
      mockItem
    </div>
    <div />
  </li>
</div>
`;

exports[`InlineList when outside of a transaction should match the snapshot on file (styles) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background: #FFFFFF;
  margin-bottom: 5px;
  padding: 5px;
  color: blue;
  min-width: 220px;
}

.c0 div:last-child {
  margin-left: auto;
}

.c0 .inline-list-item-title-div:hover {
  color: red;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 40px;
}

<div>
  <li
    class="c0"
    data-testid="ListItemElement"
  >
    <div
      class="c1 inline-list-item-title-div"
      data-testid="ListItemTitle"
    >
      mockItem
    </div>
    <div />
  </li>
</div>
`;
